source/base/data_out_base.cc

```diff
@@ -5571,11 +5593,14 @@ namespace DataOutBase
         const auto first_component = std::get<0>(range);
         const auto last_component  = std::get<1>(range);
         const auto name            = std::get<2>(range);
+        const bool is_tensor =
+          (std::get<3>(range) ==
+           DataComponentInterpretation::component_is_part_of_tensor);
         AssertThrow(last_component >= first_component,
                     ExcLowerRange(last_component, first_component));
         AssertThrow(last_component < n_data_sets,
                     ExcIndexRange(last_component, 0, n_data_sets));
-        AssertThrow(last_component + 1 - first_component <= 3,
+        AssertThrow((last_component + 1 - first_component <= 3 || is_tensor),
```

Comment:
Please do two things here:
1. conditionally check the assert using `if (!is_tensor)`
2. make the equivalent check for `if (is_tensor)`, as this should have no more than 9 components